runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -5118,172 +5118,260 @@ TR_PrexArgInfo* TR_PrexArgInfo::argInfoFromCaller(TR::Node* callNode, TR_PrexArg
+                valueNode->getOpCode().hasSymbolReference() &&
+                valueNode->getSymbolReference()->hasKnownObjectIndex())
+               {
+               prexArg = new (inliner->trHeapMemory()) TR_PrexArgument(valueNode->getSymbolReference()->getKnownObjectIndex(), comp);
+               if (tracePrex)
+                  traceMsg(comp, "PREX.inl:      %p: is known object obj%d\n", prexArg, symRef->getKnownObjectIndex());
+               }
+            }
+         }
+      else if (symRef == comp->getSymRefTab()->findJavaLangClassFromClassSymbolRef())
+         {
+         TR::Node *argFirstChild = argument->getFirstChild();
+         if (argFirstChild->getOpCode().hasSymbolReference() &&
+             argFirstChild->getSymbol()->isStatic() &&
+             !argFirstChild->getSymbolReference()->isUnresolved() &&
+             argFirstChild->getSymbol()->isClassObject() &&
+             argFirstChild->getSymbolReference()->getSymbol()->castToStaticSymbol()->getStaticAddress())
+            {
+            uintptr_t objectReferenceLocation = (uintptr_t)argFirstChild->getSymbolReference()->getSymbol()->castToStaticSymbol()->getStaticAddress();
+            TR::KnownObjectTable *knot = comp->getOrCreateKnownObjectTable();
+            if (knot)
+               {
+               TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
+               auto knownObjectIndex = knot->getOrCreateIndexAt((uintptr_t*)(objectReferenceLocation + fej9->getOffsetOfJavaLangClassFromClassField()));
+               prexArg = new (comp->trHeapMemory()) TR_PrexArgument(knownObjectIndex, comp);
+               if (tracePrex)
+                  traceMsg(comp, "PREX.inl is known java/lang/Calss obj%d\n", prexArg, knownObjectIndex);
+               }
+            }
+         }
+
+      if (prexArg)
+         {
+         if (!prexArgInfo)
+            prexArgInfo =  new (inliner->trHeapMemory()) TR_PrexArgInfo(callNode->getNumArguments(), inliner->trMemory());
+         prexArgInfo->set(argOrdinal, prexArg);
          }
       }
+
    if (tracePrex)
-      traceMsg(comp(), "PREX.inl: Done populating prex argInfo for %s %p\n", site->_callNode->getOpCode().getName(), site->_callNode);
+      traceMsg(comp, "PREX.inl: Done populating prex argInfo for %s %p\n", callNode->getOpCode().getName(), callNode);
 
-   target->_prexArgInfo = argInfo;
-   return argInfo;
+   if (tracePrex && prexArgInfo)
+      {
+      traceMsg(comp, "PREX.inl:    argInfo for callsite %p\n", site);
+      prexArgInfo->dumpTrace();
+      }
+
+   if (callerArgInfo)
+      {
+      if (tracePrex)
+         traceMsg(comp, "PREX.inl: Propagating prex argInfo from caller for [%p] %s %s\n",
+                  callNode,
+                  callNode->getOpCode().getName(),
+                  callNode->getSymbol()->castToMethodSymbol()->getMethod()->signature(inliner->trMemory(), stackAlloc));
+
+      TR_PrexArgInfo* argsFromCaller = TR_PrexArgInfo::argInfoFromCaller(callNode, callerArgInfo);
```

Comment:
I believe this is the second time this code pattern is seen wherein we effectively intesect the caller side arg info with the call target arg info. Is it worth refactoring this into a routine to make it easier to use the new logic in the future ? It is only a minor point and so it is ok if the answer is no.

Reply:
Intersecting the caller arg info with the call target/site one is already in inliner utils, it should be easy to use now providing the caller arg info. Is this the routine you meant?
```
TR_J9InlinerUtil::computePrexInfo(TR_CallTarget *target, TR_PrexArgInfo *callerArgInfo)
TR_J9InlinerUtil::computePrexInfo(TR_InlinerBase *inliner, TR_CallSite* site, TR_PrexArgInfo *callerArgInfo)
```

Reply:
No I meant the pattern where we "enhance" prex arg info based on caller and call site arg info.

Reply:
This pattern is only seen in `TR_J9InlinerUtil::computePrexInfo(TR_InlinerBase *inliner, TR_CallSite* site, TR_PrexArgInfo *callerArgInfo)`, so I'll leave it as it is now. I'll put it in a routine if it shows up elsewhere in the future.